File: ts/src/key-contract.ts
import YAML from 'yaml';

import { TheorydbError } from './errors.js';

export type KeyContractInputValue =
  | string
  | number
  | boolean
  | null
  | undefined;
export type KeyContractTransform = 'trim' | 'wildcard_empty';

export interface DerivedKeyContract {
  tabletheory_model_contract_version: '0.1';
  namespace?: string;
  dms_version?: string;
  models?: DerivedKeyModelContract[];
  derived_keys: DerivedKeyDefinition[];
}

export interface DerivedKeyModelContract {
  name: string;
  dms_model?: string;
  table?: string;
  derived_keys?: string[];
}

export interface DerivedKeyDefinition {
  name: string;
  helper?: string;
  description?: string;
  join: string;
  inputs?: DerivedKeyInput[];
  segments: DerivedKeySegment[];
  fixtures?: DerivedKeyFixture[];
}

export interface DerivedKeyInput {
  name: string;
  ts_name?: string;
  type?: 'string' | 'number' | 'boolean' | 'scalar';
  optional?: boolean;
}

export interface DerivedKeySegment {
  name?: string;
  prefix?: string;
  suffix?: string;
  literal?: string;
  value?: DerivedKeyValueSource;
  transforms?: KeyContractTransform[];
  default?: string;
  optional?: boolean;
  omit_when?: DerivedKeyOmitWhen;
}

export interface DerivedKeyValueSource {
  input?: string;
  literal?: string;
}

export interface DerivedKeyOmitWhen {
  empty?: boolean;
  default?: boolean;
  values?: string[];
}

export interface DerivedKeyFixture {
  name: string;
  description?: string;
  input: Record<string, KeyContractInputValue>;
  expect: string;
}

export function parseDerivedKeyContract(raw: string): DerivedKeyContract {
  let parsed: unknown;
  try {
    parsed = YAML.parse(raw) as unknown;
  } catch (err) {
    throw new TheorydbError(
      'ErrInvalidModel',
      'Invalid key contract YAML/JSON',
      {
        cause: err,
      },
    );
  }
  validateDerivedKeyContract(parsed);
  return parsed;
}

export function evaluateDerivedKey(
  contract: DerivedKeyContract,
  name: string,
  input: Record<string, KeyContractInputValue> = {},
): string {
  validateDerivedKeyContract(contract);
  const key = contract.derived_keys.find(
    (candidate) => candidate.name === name,
  );
  if (!key) {
    throw new TheorydbError(
      'ErrInvalidModel',
      `Derived key not found: ${name}`,
    );
  }
  return evaluateDerivedKeyDefinition(key, input);
}

export function evaluateDerivedKeyDefinition(
  key: DerivedKeyDefinition,
  input: Record<string, KeyContractInputValue> = {},
): string {
  validateDerivedKeyDefinition(key);
  const parts: string[] = [];
  for (const [i, segment] of key.segments.entries()) {
    if (segment === undefined) continue;
    const evaluated = evaluateSegment(key.name, i, segment, input);
    if (!evaluated.omit) {
      parts.push(
        `${segment.prefix ?? ''}${evaluated.value}${segment.suffix ?? ''}`,
      );
    }
  }
  return parts.join(key.join);
}

export function verifyDerivedKeyFixtures(contract: DerivedKeyContract): void {
  validateDerivedKeyContract(contract);
  for (const key of contract.derived_keys) {
    for (const fixture of key.fixtures ?? []) {
      const got = evaluateDerivedKeyDefinition(key, fixture.input);
      if (got !== fixture.expect) {
        throw new TheorydbError(
          'ErrInvalidModel',
          `Derived key ${key.name} fixture ${fixture.name}: expected ${JSON.stringify(
            fixture.expect,
          )}, got ${JSON.stringify(got)}`,
        );
      }
    }
  }
}

export function validateDerivedKeyContract(
  value: unknown,
): asserts value is DerivedKeyContract {
  if (!isPlainObject(value)) {
    throw new TheorydbError(
      'ErrInvalidModel',
      'Key contract document must be an object',
    );
  }
  if (value.tabletheory_model_contract_version !== '0.1') {
    throw new TheorydbError(
      'ErrInvalidModel',
      `Unsupported tabletheory_model_contract_version: ${String(
        value.tabletheory_model_contract_version ?? '',
      )}`,
    );
  }
  if (!Array.isArray(value.derived_keys) || value.derived_keys.length === 0) {
    throw new TheorydbError(
      'ErrInvalidModel',
      'Key contract must include derived_keys[]',
    );
  }

  const names = new Set<string>();
  for (const key of value.derived_keys) {
    validateDerivedKeyDefinition(key);
    if (names.has(key.name)) {
      throw new TheorydbError(
        'ErrInvalidModel',
        `Duplicate derived key: ${key.name}`,
      );
    }
    names.add(key.name);
  }

  if (value.models !== undefined) {
    if (!Array.isArray(value.models)) {
      throw new TheorydbError(
        'ErrInvalidModel',
        'Key contract models must be an array',
      );
    }
    for (const model of value.models) {
      if (
        !isPlainObject(model) ||
        typeof model.name !== 'string' ||
        model.name === ''
      ) {
        throw new TheorydbError(
          'ErrInvalidModel',
          'Model contract missing name',
        );
      }
      if (model.derived_keys !== undefined) {
        if (!Array.isArray(model.derived_keys)) {
          throw new TheorydbError(
            'ErrInvalidModel',
            `Model ${model.name} derived_keys must be an array`,
          );
        }
        for (const keyName of model.derived_keys) {
          if (typeof keyName !== 'string' || !names.has(keyName)) {
            throw new TheorydbError(
              'ErrInvalidModel',
              `Model ${model.name} references unknown derived key: ${String(keyName)}`,
            );
          }
        }
      }
    }
  }
}

export function validateDerivedKeyDefinition(
  value: unknown,
): asserts value is DerivedKeyDefinition {
  if (
    !isPlainObject(value) ||
    typeof value.name !== 'string' ||
    value.name === ''
  ) {
    throw new TheorydbError('ErrInvalidModel', 'Derived key missing name');
  }
  if (typeof value.join !== 'string') {
    throw new TheorydbError(
      'ErrInvalidModel',
      `Derived key ${value.name}: join must be a string`,
    );
  }
  if (!Array.isArray(value.segments) || value.segments.length === 0) {
    throw new TheorydbError(
      'ErrInvalidModel',
      `Derived key ${value.name}: missing segments[]`,
    );
  }

  const inputNames = new Set<string>();
  if (value.inputs !== undefined) {
    if (!Array.isArray(value.inputs)) {
      throw new TheorydbError(
        'ErrInvalidModel',
        `Derived key ${value.name}: inputs must be an array`,
      );
    }
    for (const input of value.inputs) {
      if (
        !isPlainObject(input) ||
        typeof input.name !== 'string' ||
        input.name === ''
      ) {
        throw new TheorydbError(
          'ErrInvalidModel',
          `Derived key ${value.name}: input missing name`,
        );
      }
      if (inputNames.has(input.name)) {
        throw new TheorydbError(
          'ErrInvalidModel',
          `Derived key ${value.name}: duplicate input ${input.name}`,
        );
      }
      inputNames.add(input.name);
    }
  }

  for (let i = 0; i < value.segments.length; i++) {
    validateSegment(value.name, i, value.segments[i], inputNames);
  }

  if (value.fixtures !== undefined) {
    if (!Array.isArray(value.fixtures)) {
      throw new TheorydbError(
        'ErrInvalidModel',
        `Derived key ${value.name}: fixtures must be an array`,
      );
    }
    const fixtureNames = new Set<string>();
    for (const fixture of value.fixtures) {
      if (
        !isPlainObject(fixture) ||
        typeof fixture.name !== 'string' ||
        fixture.name === ''
      ) {
        throw new TheorydbError(
          'ErrInvalidModel',
          `Derived key ${value.name}: fixture missing name`,
        );
      }
      if (fixtureNames.has(fixture.name)) {
        throw new TheorydbError(
          'ErrInvalidModel',
          `Derived key ${value.name}: duplicate fixture ${fixture.name}`,
        );
      }
      fixtureNames.add(fixture.name);
      if (!isPlainObject(fixture.input)) {
        throw new TheorydbError(
          'ErrInvalidModel',
          `Derived key ${value.name} fixture ${fixture.name}: missing input`,
        );
      }
      if (typeof fixture.expect !== 'string') {
        throw new TheorydbError(
          'ErrInvalidModel',
          `Derived key ${value.name} fixture ${fixture.name}: expect must be a string`,
        );
      }
    }
  }
}

function validateSegment(
  keyName: string,
  index: number,
  value: unknown,
  inputNames: ReadonlySet<string>,
): asserts value is DerivedKeySegment {
  const label =
    isPlainObject(value) && typeof value.name === 'string' && value.name !== ''
      ? value.name
      : `#${index + 1}`;
  if (!isPlainObject(value)) {
    throw new TheorydbError(
      'ErrInvalidModel',
      `Derived key ${keyName} segment ${label}: segment must be an object`,
    );
  }

  const sourceCount =
    (typeof value.literal === 'string' ? 1 : 0) +
    (isPlainObject(value.value) &&
    typeof value.value.input === 'string' &&
    value.value.input !== ''
      ? 1
      : 0) +
    (isPlainObject(value.value) && typeof value.value.literal === 'string'
      ? 1
      : 0);
  if (sourceCount !== 1) {
    throw new TheorydbError(
      'ErrInvalidModel',
      `Derived key ${keyName} segment ${label}: exactly one value source is required`,
    );
  }

  const inputName =
    isPlainObject(value.value) && typeof value.value.input === 'string'
      ? value.value.input
      : '';
  if (inputName !== '' && inputNames.size > 0 && !inputNames.has(inputName)) {
    throw new TheorydbError(
      'ErrInvalidModel',
      `Derived key ${keyName} segment ${label}: unknown input ${inputName}`,
    );
  }

  if (value.transforms !== undefined) {
    if (!Array.isArray(value.transforms)) {
      throw new TheorydbError(
        'ErrInvalidModel',
        `Derived key ${keyName} segment ${label}: transforms must be an array`,
      );
    }
    for (const transform of value.transforms) {
      if (transform !== 'trim' && transform !== 'wildcard_empty') {
        throw new TheorydbError(
          'ErrInvalidModel',
          `Derived key ${keyName} segment ${label}: unsupported transform ${String(transform)}`,
        );
      }
    }
  }
}

function evaluateSegment(
  keyName: string,
  index: number,
  segment: DerivedKeySegment,
  input: Record<string, KeyContractInputValue>,
): { value: string; omit: boolean } {
  const label =
    segment.name && segment.name !== '' ? segment.name : `#${index + 1}`;
  const resolved = resolveSegmentValue(segment, input, keyName, label);
  if (!resolved.present && segment.optional && segment.default === undefined) {
    return { value: '', omit: true };
  }

  let value = resolved.value;
  for (const transform of segment.transforms ?? []) {
    switch (transform) {
      case 'trim':
        value = value.trim();
        break;
      case 'wildcard_empty':
        if (value === '') value = '*';
        break;
      default:
        throw new TheorydbError(
          'ErrInvalidModel',
          `Derived key ${keyName} segment ${label}: unsupported transform ${String(transform)}`,
        );
    }
  }

  if (value === '' && segment.default !== undefined) {
    value = segment.default;
  }

  if (shouldOmitSegment(value, segment)) {
    return { value: '', omit: true };
  }
  if (value === '' && segment.optional && segment.default === undefined) {
    return { value: '', omit: true };
  }
  return { value, omit: false };
}

function resolveSegmentValue(
  segment: DerivedKeySegment,
  input: Record<string, KeyContractInputValue>,
  keyName: string,
  label: string,
): { value: string; present: boolean } {
  if (segment.literal !== undefined)
    return { value: segment.literal, present: true };
  if (segment.value?.literal !== undefined) {
    return { value: segment.value.literal, present: true };
  }
  const inputName = segment.value?.input;
  if (inputName) {
    if (!Object.prototype.hasOwnProperty.call(input, inputName)) {
      if (segment.default !== undefined || segment.optional) {
        return { value: '', present: false };
      }
      throw new TheorydbError(
        'ErrInvalidModel',
        `Derived key ${keyName} segment ${label}: missing required input ${JSON.stringify(inputName)}`,
      );
    }
    return {
      value: scalarToString(input[inputName], inputName),
      present: true,
    };
  }
  throw new TheorydbError(
    'ErrInvalidModel',
    `Derived key ${keyName} segment ${label}: missing value source`,
  );
}

function shouldOmitSegment(value: string, segment: DerivedKeySegment): boolean {
  if (segment.omit_when?.empty && value === '') return true;
  if (
    segment.omit_when?.default &&
    segment.default !== undefined &&
    value === segment.default
  ) {
    return true;
  }
  for (const candidate of segment.omit_when?.values ?? []) {
    if (value === candidate) return true;
  }
  return false;
}

function scalarToString(value: unknown, inputName: string): string {
  if (value === null || value === undefined) return '';
  switch (typeof value) {
    case 'string':
      return value;
    case 'number':
      if (!Number.isFinite(value)) {
        throw new TheorydbError(
          'ErrInvalidModel',
          `Derived key input ${inputName} must be a finite number`,
        );
      }
      return String(value);
    case 'boolean':
      return value ? 'true' : 'false';
    default:
      throw new TheorydbError(
        'ErrInvalidModel',
        `Derived key input ${inputName} must be a scalar`,
      );
  }
}

function isPlainObject(value: unknown): value is Record<string, unknown> {
  if (typeof value !== 'object' || value === null || Array.isArray(value))
    return false;
  const proto = Object.getPrototypeOf(value);
  return proto === Object.prototype || proto === null;
}

File: py/src/tabletheory_py/key_contract.py
from __future__ import annotations

import json
import math
from decimal import Decimal
from pathlib import Path
from typing import Any, cast

import yaml

from .errors import ValidationError

KeyContractInputValue = str | int | float | bool

TRANSFORM_TRIM = "trim"
TRANSFORM_WILDCARD_EMPTY = "wildcard_empty"
TRANSFORM_LOWERCASE = "lowercase"
TRANSFORM_URL_ENCODE = "url_encode"

SUPPORTED_CONTRACT_VERSIONS = {"0.1", "0.2"}
SUPPORTED_TRANSFORMS = {
    TRANSFORM_TRIM,
    TRANSFORM_WILDCARD_EMPTY,
    TRANSFORM_LOWERCASE,
    TRANSFORM_URL_ENCODE,
}


def parse_derived_key_contract(raw: str) -> dict[str, Any]:
    try:
        parsed = yaml.safe_load(raw)
    except Exception as err:
        raise ValidationError("invalid key contract YAML/JSON") from err
    validate_derived_key_contract(parsed)
    return cast(dict[str, Any], parsed)


def load_key_contract_file(path: str | Path) -> dict[str, Any]:
    return parse_derived_key_contract(Path(path).read_text(encoding="utf-8"))


def evaluate_derived_key(
    contract: dict[str, Any],
    name: str,
    input: dict[str, KeyContractInputValue] | None = None,
) -> str:
    validate_derived_key_contract(contract)
    for key in contract["derived_keys"]:
        if key["name"] == name:
            return evaluate_derived_key_definition(key, input or {})
    raise ValidationError(f"derived key not found: {name}")


def evaluate_derived_key_definition(
    key: dict[str, Any],
    input: dict[str, KeyContractInputValue] | None = None,
) -> str:
    validate_derived_key_definition(key)
    values = input or {}
    input_types = {
        input_def["name"]: input_def.get("type", "")
        for input_def in key.get("inputs", [])
        if isinstance(input_def, dict) and isinstance(input_def.get("name"), str)
    }
    parts: list[str] = []
    for index, segment in enumerate(key["segments"]):
        evaluated_value, omit = _evaluate_segment(key["name"], index, segment, values, input_types)
        if not omit:
            parts.append(f"{segment.get('prefix', '')}{evaluated_value}{segment.get('suffix', '')}")
    return str(key.get("join", "")).join(parts)


def verify_derived_key_fixtures(contract: dict[str, Any]) -> None:
    validate_derived_key_contract(contract)
    for key in contract["derived_keys"]:
        for fixture in key.get("fixtures", []):
            got = evaluate_derived_key_definition(key, fixture["input"])
            if got != fixture["expect"]:
                raise ValidationError(
                    f"derived key {key['name']} fixture {fixture['name']}: "
                    f"expected {json.dumps(fixture['expect'])}, got {json.dumps(got)}"
                )


def validate_derived_key_contract(value: Any) -> None:
    if not isinstance(value, dict):
        raise ValidationError("key contract document must be an object")
    version = value.get("tabletheory_model_contract_version")
    if version not in SUPPORTED_CONTRACT_VERSIONS:
        raise ValidationError(f"unsupported tabletheory_model_contract_version: {version!r}")
    derived_keys = value.get("derived_keys")
    if not isinstance(derived_keys, list) or not derived_keys:
        raise ValidationError("key contract must include derived_keys[]")

    names: set[str] = set()
    for key in derived_keys:
        validate_derived_key_definition(key)
        key_name = key["name"]
        if key_name in names:
            raise ValidationError(f"duplicate derived key: {key_name}")
        names.add(key_name)

    models = value.get("models")
    if models is None:
        return
    if not isinstance(models, list):
        raise ValidationError("key contract models must be an array")
    for model in models:
        if not isinstance(model, dict) or not isinstance(model.get("name"), str) or not model["name"]:
            raise ValidationError("model contract missing name")
        derived = model.get("derived_keys")
        if derived is None:
            continue
        if not isinstance(derived, list):
            raise ValidationError(f"model {model['name']} derived_keys must be an array")
        for key_name in derived:
            if not isinstance(key_name, str) or key_name not in names:
                raise ValidationError(f"model {model['name']} references unknown derived key: {key_name!r}")


def validate_derived_key_definition(value: Any) -> None:
    if not isinstance(value, dict) or not isinstance(value.get("name"), str) or not value["name"]:
        raise ValidationError("derived key missing name")
    if not isinstance(value.get("join"), str):
        raise ValidationError(f"derived key {value['name']}: join must be a string")
    segments = value.get("segments")
    if not isinstance(segments, list) or not segments:
        raise ValidationError(f"derived key {value['name']}: missing segments[]")

    input_names: set[str] = set()
    inputs = value.get("inputs", [])
    if inputs is None:
        inputs = []
    if not isinstance(inputs, list):
        raise ValidationError(f"derived key {value['name']}: inputs must be an array")
    for input_def in inputs:
        if (
            not isinstance(input_def, dict)
            or not isinstance(input_def.get("name"), str)
            or not input_def["name"]
        ):
            raise ValidationError(f"derived key {value['name']}: input missing name")
        input_name = input_def["name"]
        if input_name in input_names:
            raise ValidationError(f"derived key {value['name']}: duplicate input {input_name}")
        input_names.add(input_name)

    for index, segment in enumerate(segments):
        _validate_segment(value["name"], index, segment, input_names)

    fixtures = value.get("fixtures", [])
    if fixtures is None:
        fixtures = []
    if not isinstance(fixtures, list):
        raise ValidationError(f"derived key {value['name']}: fixtures must be an array")
    fixture_names: set[str] = set()
    for fixture in fixtures:
        if not isinstance(fixture, dict) or not isinstance(fixture.get("name"), str) or not fixture["name"]:
            raise ValidationError(f"derived key {value['name']}: fixture missing name")
        fixture_name = fixture["name"]
        if fixture_name in fixture_names:
            raise ValidationError(f"derived key {value['name']}: duplicate fixture {fixture_name}")
        fixture_names.add(fixture_name)
        if not isinstance(fixture.get("input"), dict):
            raise ValidationError(f"derived key {value['name']} fixture {fixture_name}: missing input")
        if not isinstance(fixture.get("expect"), str):
            raise ValidationError(
                f"derived key {value['name']} fixture {fixture_name}: expect must be a string"
            )


def _validate_segment(key_name: str, index: int, value: Any, input_names: set[str]) -> None:
    label = (
        value.get("name")
        if isinstance(value, dict) and isinstance(value.get("name"), str)
        else f"#{index + 1}"
    )
    if not isinstance(value, dict):
        raise ValidationError(f"derived key {key_name} segment {label}: segment must be an object")
    segment_value = value.get("value")
    source_count = (
        (1 if isinstance(value.get("literal"), str) else 0)
        + (
            1
            if isinstance(segment_value, dict)
            and isinstance(segment_value.get("input"), str)
            and segment_value.get("input") != ""
            else 0
        )
        + (1 if isinstance(segment_value, dict) and isinstance(segment_value.get("literal"), str) else 0)
    )
    if source_count != 1:
        raise ValidationError(f"derived key {key_name} segment {label}: exactly one value source is required")

    input_name = segment_value.get("input") if isinstance(segment_value, dict) else None
    if isinstance(input_name, str) and input_name and input_names and input_name not in input_names:
        raise ValidationError(f"derived key {key_name} segment {label}: unknown input {input_name}")

    transforms = value.get("transforms", [])
    if transforms is None:
        transforms = []
    if not isinstance(transforms, list):
        raise ValidationError(f"derived key {key_name} segment {label}: transforms must be an array")
    for transform in transforms:
        if transform not in SUPPORTED_TRANSFORMS:
            raise ValidationError(
                f"derived key {key_name} segment {label}: unsupported transform {transform!r}"
            )


def _evaluate_segment(
    key_name: str,
    index: int,
    segment: dict[str, Any],
    input_values: dict[str, KeyContractInputValue],
    input_types: dict[str, Any],
) -> tuple[str, bool]:
    label = (
        segment.get("name")
        if isinstance(segment.get("name"), str) and segment.get("name")
        else f"#{index + 1}"
    )
    value, present, from_input = _resolve_segment_value(
        segment, input_values, input_types, key_name, str(label)
    )
    if not present and segment.get("optional") is True and segment.get("default") is None:
        return "", True

    value, generated_wildcard, percent_encoded = _apply_transforms(value, segment.get("transforms") or [])
    if from_input and not generated_wildcard and not percent_encoded:
        value = _url_encode(value)
    if value == "" and segment.get("default") is not None:
        value = str(segment["default"])

    if _should_omit_segment(value, segment):
        return "", True
    if value == "" and segment.get("optional") is True and segment.get("default") is None:
        return "", True
    return value, False


def _resolve_segment_value(
    segment: dict[str, Any],
    input_values: dict[str, KeyContractInputValue],
    input_types: dict[str, Any],
    key_name: str,
    label: str,
) -> tuple[str, bool, bool]:
    if isinstance(segment.get("literal"), str):
        return segment["literal"], True, False
    value = segment.get("value")
    if isinstance(value, dict) and isinstance(value.get("literal"), str):
        return value["literal"], True, False
    if isinstance(value, dict) and isinstance(value.get("input"), str) and value["input"]:
        input_name = value["input"]
        if input_name not in input_values:
            if segment.get("default") is not None or segment.get("optional") is True:
                return "", False, False
            raise ValidationError(
                f"derived key {key_name} segment {label}: missing required input {input_name!r}"
            )
        return (
            _scalar_to_string(input_values[input_name], input_name, str(input_types.get(input_name, ""))),
            True,
            True,
        )
    raise ValidationError(f"derived key {key_name} segment {label}: missing value source")


def _apply_transforms(value: str, transforms: list[str]) -> tuple[str, bool, bool]:
    generated_wildcard = False
    percent_encoded = False
    for transform in transforms:
        if transform == TRANSFORM_TRIM:
            value = _trim_contract_whitespace(value)
        elif transform == TRANSFORM_WILDCARD_EMPTY:
            if value == "":
                value = "*"
                generated_wildcard = True
        elif transform == TRANSFORM_LOWERCASE:
            value = _ascii_lowercase(value)
        elif transform == TRANSFORM_URL_ENCODE:
            value = _url_encode(value)
            percent_encoded = True
        else:
            raise ValidationError(f"unsupported transform {transform!r}")
    return value, generated_wildcard, percent_encoded


def _should_omit_segment(value: str, segment: dict[str, Any]) -> bool:
    omit_when = segment.get("omit_when")
    if not isinstance(omit_when, dict):
        return False
    if omit_when.get("empty") is True and value == "":
        return True
    if (
        omit_when.get("default") is True
        and segment.get("default") is not None
        and value == str(segment["default"])
    ):
        return True
    values = omit_when.get("values", [])
    return isinstance(values, list) and any(value == candidate for candidate in values)


def _scalar_to_string(value: Any, input_name: str, input_type: str = "") -> str:
    if value is None:
        raise ValidationError(f"derived key input {input_name} must not be null")
    if isinstance(value, str):
        if input_type == "number":
            try:
                decimal = Decimal(value)
            except Exception as err:
                raise ValidationError(f"derived key input {input_name} must be a finite number") from err
            if not decimal.is_finite():
                raise ValidationError(f"derived key input {input_name} must be a finite number")
            if decimal == 0:
                return "0"
            return format(decimal, "f")
        return value
    if isinstance(value, bool):
        return "true" if value else "false"
    if isinstance(value, int):
        return str(value)
    if isinstance(value, float):
        if not math.isfinite(value):
            raise ValidationError(f"derived key input {input_name} must be a finite number")
        if value == 0:
            return "0"
        return _expand_exponent_decimal(str(value))
    if isinstance(value, Decimal):
        if not value.is_finite():
            raise ValidationError(f"derived key input {input_name} must be a finite number")
        if value == 0:
            return "0"
        return format(value, "f")
    raise ValidationError(f"derived key input {input_name} must be a scalar")


_UNRESERVED_BYTES = set(b"ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-._~")


def _url_encode(value: str) -> str:
    out: list[str] = []
    for byte in value.encode("utf-8"):
        if byte in _UNRESERVED_BYTES:
            out.append(chr(byte))
        else:
            out.append(f"%{byte:02X}")
    return "".join(out)


def _ascii_lowercase(value: str) -> str:
    return "".join(chr(ord(ch) + 0x20) if "A" <= ch <= "Z" else ch for ch in value)


_TRIM_CODEPOINTS = {
    0x0009,
    0x000A,
    0x000B,
    0x000C,
    0x000D,
    0x0020,
    0x0085,
    0x00A0,
    0x1680,
    0x2000,
    0x2001,
    0x2002,
    0x2003,
    0x2004,
    0x2005,
    0x2006,
    0x2007,
    0x2008,
    0x2009,
    0x200A,
    0x2028,
    0x2029,
    0x202F,
    0x205F,
    0x3000,
    0xFEFF,
}


def _trim_contract_whitespace(value: str) -> str:
    start = 0
    while start < len(value) and ord(value[start]) in _TRIM_CODEPOINTS:
        start += 1
    end = len(value)
    while end > start and ord(value[end - 1]) in _TRIM_CODEPOINTS:
        end -= 1
    return value[start:end]


def _expand_exponent_decimal(value: str) -> str:
    if "e" not in value and "E" not in value:
        return value
    return format(Decimal(value), "f")
